### ISSUE TITLE
Wrong images in validation set in custom itemlist

### ISSUE THREAD

**Issue by abhinavtripathi95** [fastai/fastai#2586]
**Describe the bug**

While trying to use a custom itemlist using [this tutorial](https://docs.fast.ai/tutorial.itemlist.html), `data.show_batch(rows, DatasetType.Train)` shows the right pair of images. However, `data.show_batch(rows, DatasetType.Valid)` shows wrong image pairs.

**Provide your installation details**
=== Software === 
python        : 3.6.9
fastai        : 1.0.61
fastprogress  : 0.2.2
torch         : 1.4.0
nvidia driver : 410.78
torch cuda    : 10.0 / is available
torch cudnn   : 7603 / is enabled

**To Reproduce**
Create a dataframe which looks like this. We use this to create our dataset.
```python
df.head()
```
|   |                                    uav |                                       sat | label | is_valid |
|--:|---------------------------------------:|------------------------------------------:|------:|---------:|
| 0 | train/atlanta/atlanta_uav/uav/uav0.png | train/atlanta/atlanta_sat/sat300/sat0.png |     1 |    False |
| 1 | train/atlanta/atlanta_uav/uav/uav1.png | train/atlanta/atlanta_sat/sat300/sat1.png |     1 |    False |
| 2 | train/atlanta/atlanta_uav/uav/uav2.png | train/atlanta/atlanta_sat/sat300/sat2.png |     1 |    False |
| 3 | train/atlanta/atlanta_uav/uav/uav3.png | train/atlanta/atlanta_sat/sat300/sat3.png |     1 |    False |
| 4 | train/atlanta/atlanta_uav/uav/uav4.png | train/atlanta/atlanta_sat/sat300/sat4.png |     1 |    False |
```python
mean, std = torch.tensor(imagenet_stats)
# for 3 channels
mean = mean[...,None,None]
std = std[...,None,None]

# Create a custom itemlist
# The primary difference from the tutorial is with how normalization is being done here
class ImageTuple(ItemBase):
    def __init__(self, img1, img2):
        self.img1,self.img2 = img1,img2
        self.obj = (img1,img2)
        self.data = [(img1.data-mean)/std, (img2.data-mean)/std]
    def apply_tfms(self, tfms_list, **kwargs):
        self.img1 = self.img1.apply_tfms(tfms_list[0], **kwargs)
        self.img2 = self.img2.apply_tfms(tfms_list[1], **kwargs)
        self.data = [(self.img1.data-mean)/std, (self.img2.data-mean)/std]
        return self    
    def to_one(self): return Image(mean+torch.cat(self.data,2)*std)
    def __repr__(self): return f'{self.__class__.__name__} {self.img1.shape, self.img2.shape}'
def _maybe_squeeze(arr): return arr if is1d(arr) else (np.array(arr).ravel() if np.array(arr).shape == () else np.squeeze(arr))

class ImageTupleList(ImageList):
    def __init__(self, items, itemsB=None, **kwargs):
        super().__init__(items, **kwargs)
        self.itemsB = itemsB
        self.copy_new.append('itemsB')
    def get(self, i):
        img1 = super().get(i)
        fn = self.itemsB[i]
        return ImageTuple(img1, open_image(fn))
    
    @classmethod
    def from_dfs(cls, df:DataFrame, path='.', cols=0, colsB=1, **kwargs):
        "Create an `ItemList` in `path` from the inputs in the `cols` of `df`."
        itemsB = ImageList.from_df(df, path, colsB).items
        res = super().from_df(df,path,cols, itemsB=itemsB, **kwargs)
        res.path = path
        return res

    def reconstruct(self, t:Tensor): 
        return ImageTuple(Image(t[0]*std+mean),Image(t[1]*std+mean))
    def show_xys(self, xs, ys, figsize:Tuple[int,int]=(12,6), **kwargs):
        rows = int(math.sqrt(len(xs)))
        fig, axs = plt.subplots(rows,rows,figsize=figsize)
        for i, ax in enumerate(axs.flatten() if rows > 1 else [axs]):
            xs[i].to_one().show(ax=ax, y=ys[i], **kwargs)
        plt.tight_layout()

    def show_xyzs(self, xs, ys, zs, figsize:Tuple[int,int]=None, **kwargs):
        """Show `xs` (inputs), `ys` (targets) and `zs` (predictions) on a figure of `figsize`.
        `kwargs` are passed to the show method."""
        figsize = ifnone(figsize, (12,3*len(xs)))
        fig,axs = plt.subplots(len(xs), 2, figsize=figsize)
        fig.suptitle('Ground truth / Predictions', weight='bold', size=14)
        for i,(x,z) in enumerate(zip(xs,zs)):
            x.to_one().show(ax=axs[i,0], **kwargs)
            z.to_one().show(ax=axs[i,1], **kwargs)


target_col = 2
is_valid_col = 3
bs = 64
data = (ImageTupleList.from_dfs(df[:-10], path='.', cols=0, colsB=1)
         .split_from_df(is_valid_col)
         .label_from_df(cols=target_col, label_cls=FloatList)
         .databunch(bs=bs))
```
So far everything works fine. If I use `show_batch` for training set, I see the correct pair of images.
```python
data.show_batch(2, DatasetType.Train)
```
However, when I use 
```python
data.show_batch(2, DatasetType.Valid)
```
it shows wrong pair of images.

**Expected behavior**

`data.show_batch(2, DatasetType.Valid)` should show pair of images in bold below
|      |                                           uav |                                              sat | label | is_valid |
|-----:|----------------------------------------------:|-------------------------------------------------:|------:|---------:|
| 1855 |     train/atlanta/atlanta_uav/uav/uav1855.png |     train/atlanta/atlanta_sat/sat300/sat1865.png |     1 |    False |
| 1856 |     train/atlanta/atlanta_uav/uav/uav1856.png |     train/atlanta/atlanta_sat/sat300/sat1856.png |     1 |    False |
| 1857 | **train/atlanta/atlanta_uav/uav/uav1857.png** | **train/atlanta/atlanta_sat/sat300/sat1857.png** |     1 |     True |
| 1858 | **train/atlanta/atlanta_uav/uav/uav1858.png** | **train/atlanta/atlanta_sat/sat300/sat1858.png** |     1 |     True |
| 1859 | **train/atlanta/atlanta_uav/uav/uav1859.png** | **train/atlanta/atlanta_sat/sat300/sat1859.png** |     1 |     True |
| 1860 | **train/atlanta/atlanta_uav/uav/uav1860.png** | **train/atlanta/atlanta_sat/sat300/sat1860.png** |     1 |     True |
| 1861 | **train/atlanta/atlanta_uav/uav/uav1861.png** | **train/atlanta/atlanta_sat/sat300/sat1861.png** |     1 |     True |

**Actual behavior**

Instead, it shows the pair of images in bold below
| uav                                           | sat                                           |
|-----------------------------------------------|-----------------------------------------------|
| **train/atlanta/atlanta_uav/uav/uav1857.png** | **train/atlanta/atlanta_sat/sat300/sat0.png** |
| **train/atlanta/atlanta_uav/uav/uav1858.png** | **train/atlanta/atlanta_sat/sat300/sat1.png** |
| **train/atlanta/atlanta_uav/uav/uav1859.png** | **train/atlanta/atlanta_sat/sat300/sat2.png** |
| **train/atlanta/atlanta_uav/uav/uav1860.png** | **train/atlanta/atlanta_sat/sat300/sat3.png** |
| **train/atlanta/atlanta_uav/uav/uav1861.png** | **train/atlanta/atlanta_sat/sat300/sat4.png** |

**Comment by abhinavtripathi95** [comment]
If it helps anyone, I found a workaround. Instead of an explicit splitting function, I made the split inside the class method `from_dfs`. Now the images in the validation set are correct (though we get a shuffled set).
```python
    @classmethod
    def from_dfs(cls, df:DataFrame, path='.', cols=0, colsB=1, **kwargs):
        "Create an `ItemList` in `path` from the inputs in the `cols` of `df`."
        itemsB = ImageList.from_df(df[df['is_valid']==False], path, colsB).items
        t_res = super().from_df(df[df['is_valid']==False],path,cols, itemsB=itemsB, **kwargs)
        t_res.path = path
        itemsB = ImageList.from_df(df[df['is_valid']==True], path, colsB).items
        v_res = super().from_df(df[df['is_valid']==True],path,cols, itemsB=itemsB, **kwargs)
        v_res.path = path
        return ItemLists(t_res.path, t_res, v_res)
```
To create databunch
```python
bs = 32
data = (ImageTupleList.from_dfs(df, path='.', cols=0, colsB=1)
         .label_from_df(cols=target_col, label_cls=FloatList)
         .databunch(bs=bs))
```

**Comment by jph00** [comment]
Thanks for posting the workaround!
